src/cbc_syslog/util/carbon_black_cloud.py

```diff
@@ -116,48 +116,54 @@ def build_query(cb, alert_criteria, start, end):
 
             return query
 
         # Iterate saved instances to fetch alerts for each instance
         for instance in self.instances:
             cb = instance["api"]
-            org_alerts = []
-
-            try:
-                # Fetch initial Alert batch
-                org_alerts.extend(build_query(cb, instance["alert_criteria"], start, end)[0:10000])
-
-                # Check if 10k limit was hit and iteratively fetch remaining alerts
-                #   by increasing start time to the last alert fetched
-                if len(org_alerts) >= 10000:
-                    last_alert = org_alerts[-1]
-                    while True:
-                        new_start = datetime.strptime(last_alert[time_field], time_format) + timedelta(milliseconds=1)
-                        overflow = build_query(cb, instance["alert_criteria"], new_start, end)
-
-                        # Extend alert list with follow up alert batches
-                        org_alerts.extend(overflow[0:10000])
-                        if len(overflow) >= 10000:
-                            last_alert = org_alerts[-1]
-                        else:
-                            break
-                alerts.extend(org_alerts)
-            except:
-                log.exception(f"Failed to fetch alerts (start: {start} - end: {end}) for org {cb.credentials.org_key}")
-                failed_orgs.append(cb.credentials.org_key)
+
+            # Perform alert fetch for each alert rule
+            for alert_rule in instance["alert_rules"]:
+                org_alerts = []
+
+                try:
+                    # Fetch initial Alert batch
+                    org_alerts.extend(build_query(cb, alert_rule, start, end)[0:10000])
+
+                    # Check if 10k limit was hit and iteratively fetch remaining alerts
+                    #   by increasing start time to the last alert fetched
+                    if len(org_alerts) >= 10000:
+                        last_alert = org_alerts[-1]
+                        while True:
+                            new_start = datetime.strptime(last_alert[time_field], time_format) + timedelta(milliseconds=1)
+                            overflow = build_query(cb, alert_rule, new_start, end)
+
+                            # Extend alert list with follow up alert batches
+                            org_alerts.extend(overflow[0:10000])
+                            if len(overflow) >= 10000:
+                                last_alert = org_alerts[-1]
+                            else:
+                                break
+                    alerts.extend(org_alerts)
+                except:
+                    log.exception(f"Failed to fetch alerts (start: {start} - end: {end})"
+                                  f" for org {cb.credentials.org_key} with rule configuration {alert_rule}")
+                    # Only add failed org_key once
+                    if cb.credentials.org_key not in failed_orgs:
```

Comment:
Maybe use a set? Although it is similar.

Reply:
I don't expect there to be many orgs or rules so this should be a fairly small iteration. Plus ideally there should not be many failed orgs